### ISSUE TITLE
Evaluating contrastive learning

### ISSUE THREAD

**Issue by baraujo98** [facebookresearch/DepthContrast#19]
This is a more of a theory question: How could I evaluate the sucess of this contrastive representation learning in an unsupervised manner, using validation data?

What I'm thinking for now:
1. Loading a checkpoint, doing a forward pass of a validation dataset, calculating the `NCELossMoco` loss, and analysing the its evolution.
2. Using t-SNE to plot the embeddings of different pointclouds after multiple random augmentations, to do a qualitative analysis.
3. Formulating a new expression similar to the NCE loss (but something simpler, not necessarily derivable), doing forward passes of the validation data and calculating the formula. I found no expressions of this kind, but I'm new to contrastive learning.

_PS._
I've actually done the t-SNE if your are interested. I used the embeddings generated by a model pretrained on the KITTI training set. Each color corresponds to a particular KITTI sample (identified by its number). Hexagons represent embeddings from PointNet++ and X's are from spconv-UNet. The augmenations done are particularly strong with multiple bigger patches removed, a more tight cuboid crop, random noise, X and Y flips.

![tSNE](https://user-images.githubusercontent.com/47788309/118381963-22740100-b5e8-11eb-8203-6f87eb0bf9c1.png)

**Comment by zaiweizhang** [comment]
We usually just evaluate the performance of the pre-trained model. We do it in two ways: 
1) Fix the pre-trained model and directly apply a linear layer on the final feature embedding and compare it with scratch model. (This is usually done with 2D contrastive learning on ImageNet pretraining)
2) Finetune the pre-trained model on downstream tasks and compare it with scratch model.

What you suggested sounds good to me. The t-SNE plots are very interesting. Thanks for showing that here. Would you mind comparing it with scratch model? 

@imisra Do you have any comments on the evaluation?

**Comment by baraujo98** [comment]
Ok, thanks for the feedback. If I understood correctly:

1. is freezing the backbone and only training a final layer, using the 128-dimensional embedding (in this case). I suppose this is only done in simpler tasks like 2D classification, is that it? I'm sure how to do it in here.
2. is what you have done in the paper.

So, this question arose to me because I noticed the success of the pretraining is only normally evaluated based on the downstream task, in a supervised fashion. But I was wondering how to evaluate the success of the pretext task (in this case contrastive learning) with no supervision: just seeing if it distinguishes pointcloud it has never seen, using no labels, and applying the same augmentation scheme.

About the plots, glad you found them interesting, I think they are too. Sure I can do more plots. But these plot points come from the 128-dimensional embeddings from the final MLP. And the scratch model does not have it. How would you suggest doing it? I'm thinking about plotting t-SNE plots in the beggining, middle and end of the pretraining, and analysing the cluster evolution.

**Comment by imisra** [comment]
This is a really interesting question. So I gather you want to evaluate your model but without using any labeled data?
I believe there isn't a right answer for this, as any unsupervised evaluation with may not correlate well with supervised evaluation schemes.
That said, I believe you can build off the contrastive criterion for evaluation using "N" validation samples.
1. For each validation sample compute A data augmentations (= N * A samples)
2. You can compute the validation NCE loss on these fixed samples.
3. You can also compute a "recall@K" metric by first computing a (N*A)^2 matrix of similarities in the embeddings
    a. For each augmentation sort the samples in decreasing order of similarities (NA - 1 similarity scores, excluding self-similarity).
    b. We want all the augmentations of the same sample (S_i) to rank higher than augmentations of other samples
    c. For a K value (1<K<=A) and compute recall, i.e., how many of the K closest embeddings are of sample S_i
    d. Compute recall@K for different K values

**Comment by baraujo98** [comment]
Thank you very much @imisra, that's what wanted to do, yes. Your suggestion makes a lot of sense, I'll definitely try to implement and give you feedback on that soon!

**Comment by baraujo98** [comment]
@imisra do you think N=100 and A=120 would be enough to draw conclusions, or not even close?
